### PR TITLE
EVG-14282 add option to filter route versions by task

### PR DESCRIPTION
**Pull request by ablack12**
I can't figure out the double lookup 😬 and from what I've gathered it would be even slower than this loop so. I gave this a max number of iterations so we don't just loop forever if their aren't a limit number of tasks.

### CODE REVIEW

**Comment by john-m-liu** [comment]
when you say double lookup, do you mean something like this? `db.versions.aggregate([{$lookup: {from:"builds", localField:"build_variants_status.build_id", foreignField:"_id", as:"b"}},{$lookup: {from:"tasks",localField:"_id",foreignField:"version", as:"t"}}]).pretty()`

I would think that this would be preferable so that you don't need to do the extra filtering until you get enough results but I don't know what you tried already

**Comment by ablack12** [comment]
This looks up every task from every build in the version, which seemed too intense. I wanted it to just look at the builds we got from the first lookup, but that wasn't quite working. Some versions have SO many tasks though; looking them all up feels like a mistake. (I guess I didn't specify that IncludeTasks should be only if a BuildId was specified though). Let me play around with it.

**Comment by ablack12** [comment]
Oh right, I'm not able to get tasks into the build variants structure; if I add
```
		if opts.IncludeTasks {
			lookupTasks := bson.M{
				"from":         task.Collection,
				"localField":   VersionIdKey,
				"foreignField": task.VersionKey,
				"as":           "build_variants.tasks",
			}
			pipeline = append(pipeline, bson.M{"$project": lookupTasks})
		}
```
the build variants just disappear so idk what to do about that.

If you want we could try meeting up for a pair program? I've spent too much time hitting my head against the wall with this; I haven't done an aggregation quite this complicated so I don't know what's possible and how. Otherwise I'd like to move forward with the iterative approach.

**Comment by john-m-liu** [comment]
ok can you try something like the following? I ran this on the data in your test, after adding version IDs to all the builds

`db.versions.aggregate([
			{
				"$match" : {
					"identifier" : "my_project",
					"r" : "gitter_request",
					"build_variants_status.build_variant" : "my_bv",
					"order" : {
						"$lt" : 19
					}
				}
			},
			{
				"$sort" : {
					"order" : -1
				}
			},
			{
				"$project" : {
                    "_id": 1,
					"create_time" : 1,
					"start_time" : 1,
					"errors" : 1,
					"order" : 1,
					"r" : 1,
					"finish_time" : 1,
					"gitspec" : 1,
					"author" : 1,
					"author_email" : 1,
					"message" : 1,
					"status" : 1,
					"build_variants_status" : {
						"$filter" : {
							"input" : "$build_variants_status",
							"as" : "item",
							"cond" : {
								"$eq" : [
									"$$item.build_variant",
									"my_bv"
								]
							}
						}
					}
				}
			},
			{
				"$lookup" : {
					"from" : "builds",
                    "let": { temp_version_id: "$_id" },
                    "as" : "build_variants",
                    "pipeline": [
                        {$match: {
                            $expr: {
                                $eq: ["$version", "$$temp_version_id"]
                            }
                        }},
                        {
                            "$lookup": {
                                "from": "tasks",
                                "let": { temp_build_id: "$_id" },
                                "as": "lookup_tasks",
                                "pipeline": [
                                    {$match: {
                                        $expr: {
                                            $and: [
                                                {$eq: ["$build_id", "$$temp_build_id"]},
                                                {$eq: ["$activated", false]},
                                            ]
                                        }
                                    }},
                                ]
                            }
                        }
                    ]
				}
			},
			{
				"$project" : {
					"create_time" : 1,
					"finish_time" : 1,
					"status" : 1,
					"message" : 1,
					"errors" : 1,
					"author" : 1,
					"order" : 1,
					"start_time" : 1,
					"build_variants" : 1,
					"gitspec" : 1
				}
			},
			{
				"$project" : {
					"build_variants.tasks" : 0
				}
			},
			{
				"$limit" : 10
			}
		]).pretty()`

File: model/version.go [hunk 1]
```diff
@@ -82,7 +82,10 @@ type Version struct {
 func (v *Version) MarshalBSON() ([]byte, error)  { return mgobson.Marshal(v) }
 func (v *Version) UnmarshalBSON(in []byte) error { return mgobson.Unmarshal(in, v) }
 
-const defaultVersionLimit = 20
+const (
+	defaultVersionLimit = 20
+	maxQueryAttempts    = 5
```
Comment (john-m-liu) [inline model/version.go:87]: this isn't used anymore 🎉

**Review by john-m-liu — APPROVED** [review]

**Pull request merged**